### ISSUE TITLE
Feature Parity with Hydra

### ISSUE THREAD

**Issue by rg9400** [Prowlarr/Prowlarr#27]
- [ ] Stats:
  - [ ] 1. Uniqueness Stat
  - [ ] 2. Successful Downloads by indexer
  - [ ] 3. Recompute stats by using a multi-select category filter up top. So it can be possible to see stats for Books or Music + Audiobooks
- [ ] Search History:
  - [ ] 1. Show ID or RSS queries
  - [ ] 2. Show categories and source in table
- [ ] Download History:
  - [ ] 1. Show success/fail by download
  - [ ] 2. Allow regrab of specific download
- [ ] Downloader:
  - [ ] 1. Give option to configure Downloader, allow sending results to this downloader
    - [ ] 	* Create option to upload NZB to downloader or to send link for manual downloads
    - [ ] 	* Need to make sure internal links can be used when sending to downloader, even if Prowlarr is being used behind RP
  - [ ] 2. Give option to configure blackhole for torrents
  - [ ] 3. Handle magnet links for Torrents
  - [ ] 4. Update status of downloads from Downloader
  - [ ] 5. Potentially show download queue? Hydra does this, but seems not super important
  - [ ] 6. Give default category-wise options for manual downloads, with option to override
- [ ] Categories:
  - [ ] 1. Create custom categories
  - [ ] 2. Override search types
  - [ ] 3. Allow restrictions to be applied for API or manual searches or both
  - [ ] 4. Category-specific restrictions such as required words, regex, forbidden words, forbidden regex, size restrictions
  - [ ] 5. Ignore results
- [ ] Statuses:
  - [ ] 1. Show indexer API hits and downloads compared to limits
  - [ ] 2. Show current state, last error, and if disabled, when it was disabled and until when
- [ ] Searching:
  - [ ] 1. Timeout
  - [ ] 2. User Agent
  - [ ] 3. Map external user agents
  - [ ] 4. Ignore load limiting/errors
  - [ ] 5. Handle Torznab categories or ignore them
  - [ ] 6. Map indexer specific categories to internal categories
  - [ ] 7. Convert Media IDs for indexers (TMDB to IMDB for example)
  - [ ] 8. Generate queries when indexer does not support IDs
  - [ ] 9. Fallback to generated queries if no results found
  - [ ] 10. Global restrictions (word filters, forbidden posters, max results age, min seeders, ignore passworded releases)
  - [ ] 11. Option to remove trailing garbage (obfuscated, scrambled, postbot)
  - [ ] 12. Use original categories from indexer or ones configured internally
  - [ ] 13. Dupe detection
  - [ ] 14. Cache results
- [ ] Manual Searching:
  - [ ] 1. Ad-hoc filters for search type, age, size, indexer
  - [ ] 2. Allow pre-selection of indexers, with easy options to invert, select all usenet, select all torrent, select all
  - [ ] 3. Allow ID based searches for relevant search types
  - [ ] 4. Pull NFO/covers from supported indexers
  - [ ] 5. Color code results so it is easy to visually correlate which indexer it is vs having to read from a bunch of data
  - [ ] 6. Configure how many results to display per page
  - [ ] 7. Configure how many results to pull intially, with option to page through more as needed (Hydra gives option to load all results or next 100, but no way to load say a user-defined amount, e.g. next 500).
  - [ ] 8. Show duplicate display triggers
  - [ ] 9. Group TV results by season/episode
  - [ ] 10. Show already downloaded indicator, give option to hide already downloaded results
  - [ ] 11. Multi-select, download as ZIP or send to nzbget
  - [ ] 12. Ad-hoc filter and sort results by title, indexer, category, size, number of hits, age
  - [ ] 13. Link to indexer, manual download, or directly send to downloader.
- [ ] General Settings:
  - [ ] 1. Configure how long to keep history/stats
  - [ ] 2. Dereferer for indexer access
  - [ ] 3. Disable SSL/SNI for specific hosts
- [ ] Indexer:
  - [ ] 1. Option to proxy or redirect for API downloads, maybe by indexer?
  - [ ] 2. VIP Expiry Date, including Lifetime option
  - [ ] 3. API and Download limits by indexer
  - [ ] 4. Scheduler for Indexers (I don't use this)
  - [ ] 5. Timeout override by indexer (I don't use this)
  - [ ] 6. Hit reset time, with option for rolling resets
  - [ ] 7. Load limiting
  - [ ] 8. User Agent Override
  - [ ] 9. Custom Parameters
  - [ ] 10. Preselect for manual searches
  - [ ] 11. Enable for API or manual searches or both
  - [ ] 12. Color for visualization in manual searches
  - [ ] 13. Overrides for cats, search IDs, search types, caps
- [ ] Notifications:
  - [ ] 1. VIP Expiry notification

**Comment by XxUnkn0wnxX** [comment]
hope Color code results get added, I miss this.
![Screen Shot 2021-07-01 at 4 32 06 pm](https://user-images.githubusercontent.com/2242880/124076958-443b2180-da8a-11eb-9c7a-3f5d8b752797.jpg)

**Comment by dephcon** [comment]
I'd really like to see the indexer uniqueness rating (number of unique downloads per indexer) and use that data long with the other stats to provide a indexer 'score' per hydra.  It's really helpful to weed out indexers that aren't pulling their weight come renewal time.

**Comment by jasonla** [comment]
+1 to `Option to remove trailing garbage`.

**Comment by batuhankrmn** [comment]
> * 12. Ad-hoc filter and sort results by title, indexer, category, size, number of hits, age

I believe this doesn't cover the #428 because it's not exactly dynamic filtering. We can set presets as custom filters but we can't filter the resulty instantly for a keyword. Let's say we checked the results and noticed there are multiple uploaders for certain movie release, so we want to filter the results by uploaders name. We have to create custom filter just for this case, we can't filter the results by just typing uploader name. (Jackett filtering model). For this reason, I do not think that the closed request #428 has been fully fulfilled. @the-lazy-fox @bakerboy448

**Comment by bakerboy448** [comment]
#428 was closed as a duplicate as this GHI covers it. Nothing indicates it or the quoted item on this GHI were completed

**Comment by JediMaster25** [comment]
Could you add advanced search to filter torrents?

Some filters I have in mind are
- search only title, all content or all content and title.
- exact match: `"Keyword1 keyword2"`
- exclude: `-keyword`
- or: `"|"` in between keywords.
- `*` parameter to match anything. Like some web series `s1ep*`

Greater than, less than, or a combination of both for:
- size
- date (age). 
- number of files
- number of seeders

**Comment by bakerboy448** [comment]
[NZBHydra2](https://github.com/theotherp/nzbhydra2)

**Comment by Enlinks** [comment]
+1 for caching, at least having the option to enable it.

It would help with repeated queries due to failures and help keep some of the load off the indexers.

**Comment by ChrisWeiss** [comment]
I would very much like to see at least 'exact string' search support. I'm pretty sure this would have to be an additional filtering step added in the client after the search results are received as different indexers are not as strict on search results.

**Comment by jonandeva** [comment]
Another person here who would love to be able to type to filter results. Saving filters and turning them on and off is clunky. This is the only thing I really miss from Jackett.

**Comment by bakerboy448** [comment]
> Is caching results planned? its hugely beneficial if you can cache the results for a certain period.

No

**Comment by Panzer1119** [comment]
> > Is caching results planned? its hugely beneficial if you can cache the results for a certain period.
> 
> 
> 
> No. Low benefit given it'd need to be the exact same query

Why? Is it so unlikely to trigger the same query?

One case that comes to my mind would be e.g. you run two Radarr/Sonarr instances, one for HD and 4K stuff, and you manually search for an episode and then you also want to manually search for the same episode on the other instance.

Caching that could help reduce waiting time and duplicate/multiple calls to the indexer.

**Comment by Qstick** [comment]
Most people have different indexer categories setup for their 4k instances, thus the search is not the same and the cache doesn't help in that scenario.

**Comment by Enlinks** [comment]
> > Is caching results planned? its hugely beneficial if you can cache the results for a certain period.
> 
> No. Low benefit given it'd need to be the exact same query

This would also be beneficial for situations where the first result failed to download.

**Comment by Enlinks** [comment]
> Most people have different indexer categories setup for their 4k instances, thus the search is not the same and the cache doesn't help in that scenario.

I use the same indexer for everything.

Only exception would be anime. Even that I use the same indexer but I have another indexer that I prioritize over that one.

**Comment by bakerboy448** [comment]
Not sure what using the same indexer has to do with different categories selected for said indexer in each app?

**Comment by westlandnigel** [comment]
Please add the 'lifetime' option. I know you could do blank, but I want to differentiate between lifetime and free (blank).

**Comment by Archmonger** [comment]
+1 to `Overrides for cats, search IDs, search types, caps`.

Switched over from Hydra and remembered how awful some of my indexer's built-in ID matchers are.

After a few days of Prowlarr use, this seems to be the only "major" missing feature with no workaround.

**Comment by barrelltitor** [comment]
Basically every issue about caching gets closed with
>why would you want to do this

**Comment by csmarshall** [comment]
Following the contributing guide's "comment so work is not duplicated" — I'd like to pick up **Searching → 2. User Agent**.

I have this working locally and wanted to check direction before opening a PR, since a few of the decisions are expensive to redo in review.

**Shape.** Two layers, precedence *per-indexer → global → built-in default*:

- a global User-Agent under Settings → Indexers, applied to all indexer requests
- a per-indexer override on the indexer edit form (advanced); blank inherits the global

It's applied in `IndexerHttpClient.ExecuteProxied/ExecuteProxiedAsync`, which covers search, login, captcha, download and capabilities in one place. The value lives on `IndexerDefinition` next to `Priority`/`Redirect`/`DownloadClientId` rather than in the per-indexer settings contracts, so it works for every implementation without renumbering any `FieldDefinition` orders. Migration, translation keys and unit tests included.

**Three decisions I'd like a call on:**

1. **Ordering vs FlareSolverr.** The UA is applied before the indexer proxies run, so a configured UA wins over the per-host UA FlareSolverr caches — it only injects that when the request has none. FlareSolverr still overrides on requests it actually solves, which seems right — the testing posted in #2646 found that a `cf_clearance` cookie only works when replayed with the UA that earned it. That's their result and I haven't reproduced it, so if you know otherwise, this is the decision it would change. Easy to flip if you'd rather the solver always win.

2. **Overriding a definition's UA.** A user-set value takes precedence over a `User-Agent` baked into a Cardigann definition's `headers:`. That's deliberate — it's the escape hatch when a definition's UA goes stale and you'd otherwise wait on a definitions PR — but it does mean someone can break rustorka/mojblink/etc. for themselves.

3. **Freeform vs presets.** Freeform textbox only, consistent with the existing IPTorrents and PixelHD fields. I'd argue against shipping a dropdown of "common browser" UAs: they go stale every few weeks, and a stale browser UA is a worse signal than Prowlarr's honest one.

**On the obvious objection.** From #658: *"pissing off an indexer that has explicitly made a point of blocking us is not usually the answer."* I'm not trying to relitigate that, and this isn't aimed at sites that have deliberately blocked Prowlarr.

It's worth separating two things, though. The concrete case already in the codebase is session consistency: IPTorrents and PixelHD both ship a "Cookie User-Agent" field — *"User-Agent associated with cookie used from Browser"* — and the `info_useragent` helper tells users to copy the UA from the same devtools panel they copied the cookie from. Generalising that is most of the value here, and today it only exists for the two indexers that happened to get a bespoke field. Beyond that, operators have their own reasons for wanting control over what their client announces, and I'd rather expose the setting than adjudicate each one.

Happy to split this into two PRs (per-indexer first, global second) if that suits the one-feature-per-PR rule better.

*Disclosure: I used an AI assistant on both the implementation and this comment. Every factual claim above is checked against the code in the repo rather than taken from the model, except where I've flagged otherwise.*
